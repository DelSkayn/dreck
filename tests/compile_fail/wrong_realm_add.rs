use dreck::*;

pub struct Container<'gc, 'own>(Option<Gc<'gc, 'own, Container<'gc, 'own>>>);

unsafe impl<'gc, 'own> Trace<'own> for Container<'gc, 'own> {
    type Gc<'to> = Container<'to, 'own>;

    fn needs_trace() -> bool
    where
        Self: Sized,
    {
        true
    }

    fn trace(&self, marker: Marker<'own, '_>) {
        self.0.trace(marker)
    }
}

fn main() {
    dreck!(owner1, arena1);
    dreck!(_owner2, arena2);

    let mut container = Container(None);
    let ptr = arena1.add(container);
    container = Container(Some(ptr));
    let ptr = arena2.add(container);

    arena1.collect(&owner1);
}
