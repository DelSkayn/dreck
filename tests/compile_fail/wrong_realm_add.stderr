warning: unused variable: `ptr`
  --> tests/compile_fail/wrong_realm_add.rs:27:9
   |
27 |     let ptr = arena2.add(container);
   |         ^^^ help: if this is intentional, prefix it with an underscore: `_ptr`
   |
   = note: `#[warn(unused_variables)]` on by default

error[E0597]: `_pin` does not live long enough
  --> tests/compile_fail/wrong_realm_add.rs:22:5
   |
22 |     dreck!(_owner2, arena2);
   |     ^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     borrowed value does not live long enough
   |     binding `_pin` declared here
...
30 | }
   | -
   | |
   | `_pin` dropped here while still borrowed
   | borrow might be used here, when `_lifetime_constrainer` is dropped and runs the `Drop` code for type `main::KeepTillScopeDrop`
   |
   = note: values in a scope are dropped in the opposite order they are defined
   = note: this error originates in the macro `dreck` (in Nightly builds, run with -Z macro-backtrace for more info)
